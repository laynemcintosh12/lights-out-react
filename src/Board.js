import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows, ncols, chanceLightStartsOn }) {
  const [board, setBoard] = useState(createBoard());

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {
    let initialBoard = [];
    // create the rows
    for (let i = 0; i < nrows; i++) {
      let row = [];
      // create the columns inside the rows
      for (let j = 0; j < ncols; j++) {
        row[j] = Math.random() < chanceLightStartsOn;
      }
      initialBoard.push(row);
    }
    return initialBoard;
  }

  function hasWon() {
    // Check if all values in the board are false
    return board.every(row => row.every(cell => !cell));
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);
  
      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on the board, flip it
        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };
  
      // Make a (deep) copy of the oldBoard
      const boardCopy = oldBoard.map(row => [...row]);
  
      // Flip the selected cell and its neighbors
      flipCell(y, x, boardCopy);
      flipCell(y - 1, x, boardCopy); // top
      flipCell(y + 1, x, boardCopy); // bottom
      flipCell(y, x - 1, boardCopy); // left
      flipCell(y, x + 1, boardCopy); // right
  
      return boardCopy;
    });
  }

    return (
    <div className="Board">
        <h1>Light Game</h1>
        {hasWon() ? <p>You Won!</p> :
        <div className="Board-Game">
            <table><tbody>
            {board.map((row, rowIndex) => (
                <tr key={rowIndex}>
                {row.map((cell, cellIndex) => (
                    <Cell
                    key={cellIndex}
                    flipCellsAroundMe={() => flipCellsAround(rowIndex + "-" + cellIndex)}
                    isLit={cell}
                    />
                ))}
                </tr>
            ))}
            </tbody></table>
        </div>
        }
    </div>
    );
}


export default Board;
