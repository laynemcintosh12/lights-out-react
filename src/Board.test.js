import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Board from "./Board";

describe("Board Component", function () {

  describe("Rendering", function () {

    it("should render without crashing", function () {
      render(<Board />);
    });

    it("should match the snapshot for a full board", function () {
      const { asFragment } = render(<Board chanceLightStartsOn={1} />);
      expect(asFragment()).toMatchSnapshot();
    });

    it("should display win state when lights are out", function () {
      const { getByText } = render(<Board chanceLightStartsOn={0} />);
      expect(getByText("You Won!")).toBeInTheDocument();
    });
  });


  describe("Cell Click", function () {

    it("should toggle lights correctly", function () {
      const { getAllByRole } = render(
        <Board nrows={3} ncols={3} chanceLightStartsOn={1} />,
      );
      const cells = getAllByRole("cell");

      // All cells start out as lit
      cells.forEach(cell => {
        expect(cell).toHaveClass("Cell-lit");
      });

      // Click on the middle cell
      fireEvent.click(cells[4]);

      // Now only cells at the corners should be lit
      let litIndices = [0, 2, 6, 8];
      cells.forEach((cell, idx) => {
        if (litIndices.includes(idx)) {
          expect(cell).toHaveClass("Cell-lit");
        } else {
          expect(cell).not.toHaveClass("Cell-lit");
        }
      });
    });

    it("should indicate winning state when clicking the board", function () {
      // Create a board that can be completed in one click
      const { queryByText, getAllByRole } = render(
        <Board nrows={1} ncols={3} chanceLightStartsOn={1} />,
      );

      expect(queryByText("You Won!")).not.toBeInTheDocument();

      const cells = getAllByRole("cell");
      fireEvent.click(cells[1]);

      expect(queryByText("You Won!")).toBeInTheDocument();
    });
  });
});
