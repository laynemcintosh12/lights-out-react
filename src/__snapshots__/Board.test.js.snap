// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board Component Rendering should match the snapshot for a full board 1`] = `
<DocumentFragment>
  <div
    class="Board"
  >
    <h1>
      Light Game
    </h1>
    <p>
      You Won!
    </p>
  </div>
</DocumentFragment>
`;
